llvm/lib/CodeGen/ExpandVectorPredication.cpp

```diff
@@ -581,6 +581,14 @@ bool CachingVPExpander::expandPredication(VPIntrinsic &VPI) {
     replaceOperation(*NewNegOp, VPI);
     return NewNegOp;
   }
+  case Intrinsic::vp_select:
+  case Intrinsic::vp_merge: {
+    assert(maySpeculateLanes(VPI) || VPI.canIgnoreVectorLengthParam());
```

Comment:
Subject to the assert, the code here appears correct, but what ensures the second clause holds?  For vp.select, we can convert undef elements to either true or false, but for vp.merge we need the pivot semantics if EVL!=VLMAX.  How do we know that EVL=VLMAX in this path?

From the test diffs below, it looks like we do handle the pivot, I'm just confused about the code structure.  Give me a pointer?

Reply:
I think the code here https://github.com/llvm/llvm-project/blob/27719745e3b0f7bed738a67498e751ca6d45bcdd/llvm/lib/CodeGen/ExpandVectorPredication.cpp#L642 forces the EVLStrategy to Convert if the intrinsic is to be converted.  That causes this code https://github.com/llvm/llvm-project/blob/27719745e3b0f7bed738a67498e751ca6d45bcdd/llvm/lib/CodeGen/ExpandVectorPredication.cpp#L679 to fold the EVL into the mask.